subprojects/core/src/main/java/org/gradle/cache/internal/VersionSpecificCacheCleanupAction.java

```diff
@@ -80,13 +83,20 @@ public boolean execute(@Nonnull CleanupProgressMonitor progressMonitor) {
     }
 
     private boolean requiresCleanup() {
+        long lastCleanupTimestamp;
+
         File gcFile = getGcFile();
         if (!gcFile.exists()) {
-            return gcFile.getParentFile().exists();
+            if (!gcFile.getParentFile().exists()) {
+                return false;
+            } else {
+                lastCleanupTimestamp = -1;
+            }
+        } else {
+            lastCleanupTimestamp = gcFile.lastModified();
         }
-        long duration = System.currentTimeMillis() - gcFile.lastModified();
-        long timeInHours = TimeUnit.MILLISECONDS.toHours(duration);
-        return timeInHours >= CLEANUP_INTERVAL_IN_HOURS;
+
+        return cleanupFrequency.requiresCleanup(lastCleanupTimestamp);
```

Comment:
One weird thing about doing it this way is that cleanupFrequency.requiresCleanup(...) doesn't treat `-1` as some special value, so the interval becomes (current time + 1) converted into hours, which happens to always be greater than the default 24 hours. -1 isn't a valid time, so maybe 0 would be better here. 

It might be simpler to implement different CleanupFrequency's if we still do the interval calculation here and pass a "n-hours since last cleanup" instead of the raw timestamp. In the case of an existing directory without a gcFile, we could report Integer.MAX_VALUE hours since last cleanup. 

I don't know if this is necessarily better and it doesn't need to be fixed to merge this.

Reply:
Yeah, this could be better.  I'll think on it some and make it cleaner.